### PR TITLE
feature 优化toposerver中model/association相关结构与代码逻辑

### PR DESCRIPTION
**Pull request by chalice-1831**
### 优化的功能:
- toposerver中model/association相关结构与代码优化

issue #5450

### CODE REVIEW

File: src/scene_server/topo_server/logics/model/association.go [hunk 1]
```diff
@@ -0,0 +1,508 @@
...
+	if !rspAsst.Result {
+		blog.Errorf("create object association failed, param: %#v , err: %s, rid: %s", data, rspAsst.ErrMsg, kit.Rid)
+		return nil, kit.CCError.New(rspAsst.Code, rspAsst.ErrMsg)
+	}
+
+	data.ID = int64(rspAsst.Data.Created.ID)
+	return data, nil
+}
+
+func (assoc *association) DeleteAssociation(kit *rest.Kit, cond mapstr.MapStr) error {
```
Comment (wcy00000000000000) [inline src/scene_server/topo_server/logics/model/association.go:270]: 去掉这个函数，这个不需要被使用，只需要DeleteAssociationWithPreCheck就可以了。然后DeleteAssociationWithPreCheck调这个函数又查了一次模型关联做校验，这个pre判断需要在DeleteAssociationWithPreCheck里加

File: src/scene_server/topo_server/logics/model/association.go [hunk 2]
```diff
@@ -0,0 +1,508 @@
...
+	}
+
+	if len(rsp.Data.Info) != 0 {
+		// object association has already been instantiated, association can not be deleted.
+		blog.Errorf("search association by associationID[%d], got instances, rid: %s", associationID, kit.Rid)
+		return kit.CCError.CCError(common.CCErrTopoAssociationHasAlreadyBeenInstantiated)
+	}
+
+	// TODO: check association on_delete action before really delete this association.
+	// all the pre check has finished, delete the association now.
```
Comment (wcy00000000000000) [inline src/scene_server/topo_server/logics/model/association.go:368]: 需要确定下on_delete是否需要支持

File: src/scene_server/topo_server/logics/model/association.go [hunk 3]
```diff
@@ -0,0 +1,508 @@
...
+		asso = append(asso, metadata.AssociationDetail{
+			AssociationKindID: association.AsstKindID,
+			Associations:      []metadata.Association{association},
+		})
+	}
+
+	return &metadata.AssociationList{Associations: asso}, nil
+}
+
+func (assoc *association) isObjectInAssocValid(kit *rest.Kit, objectID, asstObjectID string) error {
```
Comment (wcy00000000000000) [inline src/scene_server/topo_server/logics/model/association.go:468]: 可以用in这两个模型ID的方式一次性查出来

**Review by wcy00000000000000 — CHANGES_REQUESTED** [review]

**Comment by tencent-adm** [comment]
[CLA assistant check](https://cla.opensource.tencent.com/Tencent/bk-cmdb?pullRequest=5500) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.opensource.tencent.com/Tencent/bk-cmdb?pullRequest=5500) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.opensource.tencent.com/check/Tencent/bk-cmdb?pullRequest=5500) it.</sub>

**Comment by tencent-adm** [comment]
[CLA assistant check](https://cla.opensource.tencent.com/Tencent/bk-cmdb?pullRequest=5500) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.opensource.tencent.com/Tencent/bk-cmdb?pullRequest=5500) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.opensource.tencent.com/check/Tencent/bk-cmdb?pullRequest=5500) it.</sub>

File: src/scene_server/topo_server/logics/model/association.go [hunk 4]
```diff
@@ -0,0 +1,507 @@
...
+	rsp, err := assoc.SearchObjectAssociation(kit, cond)
+	if err != nil {
+		blog.Errorf("search object association failed, condition: %#v, err: %v, rid: %s", cond, err, kit.Rid)
+		return nil, err
+	}
+
+	return rsp, nil
+}
+
+func (assoc *association) SearchObjectAssociation(kit *rest.Kit,
```
Comment (wcy00000000000000) [inline src/scene_server/topo_server/logics/model/association.go:214]: 合了api machinery之后可以删掉

File: src/scene_server/topo_server/logics/model/association.go [hunk 5]
```diff
@@ -0,0 +1,507 @@
...
+
+	if err = rsp.CCError(); err != nil {
+		blog.Errorf("failed to create the association (%#v) , err: %v, rid: %s", cond, err, kit.Rid)
+		return nil, err
+	}
+	if len(rsp.Data.Info) > 0 {
+		blog.Errorf("failed to create the association (%#v), the associations %s->%s already exist, rid: %s",
+			cond, data.ObjectID, data.AsstObjID, kit.Rid)
+		return nil, kit.CCError.CCErrorf(common.CCErrTopoAssociationAlreadyExist, data.ObjectID, data.AsstObjID)
+	}
```
Comment (wcy00000000000000) [inline src/scene_server/topo_server/logics/model/association.go:278]: 是不是可以用count？

File: src/scene_server/topo_server/logics/model/association.go [hunk 6]
```diff
@@ -0,0 +1,507 @@
...
+	if err = rsp.CCError(); err != nil {
+		blog.Errorf("search association info failed, query: %#v, err: %v, rid: %s", queryCond, err, kit.Rid)
+		return err
+	}
+
+	if len(rsp.Data.Info) != 0 {
+		// object association has already been instantiated, association can not be deleted.
+		blog.Errorf("search association by associationID[%d], got instances, rid: %s", associationID, kit.Rid)
+		return kit.CCError.CCError(common.CCErrTopoAssociationHasAlreadyBeenInstantiated)
+	}
```
Comment (wcy00000000000000) [inline src/scene_server/topo_server/logics/model/association.go:361]: 是否可以用count？

File: src/scene_server/topo_server/logics/model/association.go [hunk 7]
```diff
@@ -0,0 +1,507 @@
...
+		return err
+	}
+
+	return nil
+}
+
+func (assoc *association) UpdateObjectAssociation(kit *rest.Kit, data mapstr.MapStr, assoID int64) error {
+
+	asst := new(metadata.Association)
+	err := data.MarshalJSONInto(asst)
```
Comment (wcy00000000000000) [inline src/scene_server/topo_server/logics/model/association.go:383]: data是不是可以在外层解析http请求的DecodeInto里直接解成metadata.Association类型？

**Review by wcy00000000000000 — CHANGES_REQUESTED** [review]

File: src/scene_server/topo_server/logics/model/association.go [hunk 8]
```diff
@@ -0,0 +1,507 @@
...
+		asso = append(asso, metadata.AssociationDetail{
+			AssociationKindID: association.AsstKindID,
+			Associations:      []metadata.Association{association},
+		})
+	}
+
+	return &metadata.AssociationList{Associations: asso}, nil
+}
+
+func (assoc *association) isObjectInAssocValid(kit *rest.Kit, objectID, asstObjectID string) error {
```
Comment (wcy00000000000000) [inline src/scene_server/topo_server/logics/model/association.go:476]: 这里如果改成用for循环判断出来哪个不存在是否更容易理解一点？

**Review by breezelxp — APPROVED** [review]
LGTM

**Pull request merged**